src/main/java/org/javarosa/core/util/externalizable/ExtWrapMultiMap.java

```diff
@@ -0,0 +1,120 @@
+package org.javarosa.core.util.externalizable;
+
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.Multimap;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.util.AbstractMap;
+import java.util.Collection;
+import java.util.Iterator;
+
+public class ExtWrapMultiMap extends ExternalizableWrapper {
+
+    private ExternalizableWrapper keyType;
+    private Class<? extends Multimap> mapImplementation;
+
+    /* Constructors for serialization */
+
+    public ExtWrapMultiMap(Multimap val) {
+        this(val, null);
+    }
+
+    public ExtWrapMultiMap(Multimap val, ExternalizableWrapper keyType) {
+        if (val == null) {
+            throw new NullPointerException();
+        }
+
+        this.val = val;
+        this.keyType = keyType;
+        this.mapImplementation = val.getClass();
+    }
+
+    /* Constructors for deserialization */
+
+    public ExtWrapMultiMap() {
+    }
+
+
+    public ExtWrapMultiMap(Class keyType) {
+        this(keyType, ArrayListMultimap.class);
+    }
+
+    public ExtWrapMultiMap(Class keyType, Class mapImplementation) {
+        this(new ExtWrapBase(keyType), mapImplementation);
+    }
+
+    public ExtWrapMultiMap(ExternalizableWrapper keyType, Class mapImplementation) {
+        if (keyType == null) {
+            throw new NullPointerException();
+        }
+
+        this.keyType = keyType;
+        this.mapImplementation = mapImplementation;
+    }
+
+    @Override
+    public ExternalizableWrapper clone(Object val) {
+        return new ExtWrapMultiMap((Multimap)val, keyType);
+    }
+
+    @Override
+    public void readExternal(DataInputStream in, PrototypeFactory pf) throws IOException, DeserializationException {
+        long size = ExtUtil.readNumeric(in);
+        try {
+            Multimap<Object, Object> multimap;
+            if (mapImplementation.equals(ArrayListMultimap.class)) {
```

Comment:
Sorry, to eliminate confusion: I don't think you need to retain the multimap implementation, here. The output of the ExtWrapMultiMap class can always be an ArrayListMultimap, and I think the added complexity for reflecting it here isn't worth it (and adds dangerous assumptions like the existence of a create method).

My only concern in the previous review was that in the *write* method, this wrapper was doing a blind cast to ArrayListMultimap, which seemed dangerous. It looks like you were able to change that by just relying on the generic type methods, though.

Reply:
Thanks, Did the revert. I mostly was trying to copy the existing pattern from [ExtWrapList](https://github.com/dimagi/commcare-core/blob/master/src/main/java/org/javarosa/core/util/externalizable/ExtWrapList.java#L74), but your concerns does sound valid.